p2p-specs/p2p-interface.md

```diff
@@ -0,0 +1,229 @@
+|----------------------------------|---------------------------|
+| `UserOperationWithEntryPoint`    | ``                        |
+| `NewPooledUserOperationsHashes`  | ``                        |
+
+
+Bundlers MUST reject (fail validation) messages containing an incorrect type, or invalid payload.
+
+When processing incoming gossip, Bundlers MAY de-score or disconnect peers who fail to observe these constraints.
+
+For any optional queueing, Bundlers SHOULD maintain maximum queue sizes to avoid DoS vectors.
+
+
+#### Global topics
+
+There are two primary global topics used to propagate user operations (`UserOperationWithEntryPoint`)
+and aggregate user operation hashes (`NewPooledUserOperationsHashes`) to all nodes on the network.
+
+##### `UserOperationWithEntryPoint`
+
+The `UserOperationWithEntryPoint` topic is the concatenation of EntryPoint address and UserOperation message serialized using SSZ
+
+##### `NewPooledUserOperationsHashes`
+
+The `NewPooledUserOperationsHashes` topic is used solely for propagating to all the connected nodes on the networks. One or more UserOps that have appeared in the network and which have not yet been included in a block are propagated to a fraction of the nodes connected to the network.
+
+
+##### `GetPooledUserOps`
+
+The `GetPooledUserOps` requests UserOps from the recipients UserOp mempool for a given EntryPoint contract address. The recommended soft limit for GetPooledUserOps requests is 256 hashes (8 KiB). The recipient may enforce an arbitrary limit on the response (size or serving time), which must not be considered a protocol violation.
+
+##### `PooledUserOperations`
+
+The `PooledUserOperations` is a response to the `GetPooledUserOps`, returning the requested UserOperations from the local pool. The items in the list are UserOps in the format described in ERC4337 specification. 
+
+## Constants
+
+| Name                             | Message Type              |
+|----------------------------------|---------------------------|
+| `MAX_OPS_PER_REQUEST`            | `uint64(2**10)` = 1024    |
+
+
+## Container Specifications
+
+The following types are SimpleSerialize (SSZ) containers.
+
+#### `UserOp`
+
+```python
+class UserOp(Container):
+    sender: Address
+    nonce: uint256
+    init_code: bytes
+    call_data: bytes
+    call_gas_limit: uint256
+    verification_gas_limit: uint256
+    pre_verification_gas: uint256
+    max_fee_per_gas: uint256
+    max_priority_fee_per_gas: uint256
+    paymaster_and_data: bytes
+    signature: bytes
+```
+
+#### `UserOperationWithEntryPoint`
+
+```python
+class UserOperationWithEntryPoint(Container):
+    entry_point_contract: Address
+    verified_at_block_hash: uint256
+    chain_id: uint256
+    user_operations: List[UserOp, MAX_OPS_PER_REQUEST]
+```
+
+#### `GetPooledUserOps`
+
+```python
+class GetPooledUserOps(Container):
+    request_id: uint256
+    chain_id: uint256
+    entry_point_address: List[Address, MAX_OPS_PER_REQUEST]
```

Comment:
why should a mempool request contain the entrypoint addresses?
the list of valid entrypoints is part of the mempool "spec" (as defined by the mempool-id)
the chain_id is not strictly required, but it is good nonetheless, to detect configuration error.s

Reply:
Shouldn't we also have a list of supported mempool-ids in this request, to ensure that the bundler only receives UserOps it would consider valid?

Reply:
> why should a mempool request contain the entrypoint addresses? the list of valid entrypoints is part of the mempool "spec" (as defined by the mempool-id) the chain_id is not strictly required, but it is good nonetheless, to detect configuration error.s

You are right. I'll replace entry_point_address with the list of mempool-id.